### PR TITLE
[REVIEW] Remove `bool8`

### PR DESCRIPTION
**Pull request by codereport**
Remove `bool8` and replace with `bool`

Issue: https://github.com/rapidsai/cudf/issues/3484

### CODE REVIEW

**Comment by jrhemstad** [comment]
I think we should maintain the `BOOL8` enum name on the chance that we add a 1 bit per bool type in the future to match Arrow (which we can call `BOOL1`).

**Comment by codereport** [comment]
> I think we should maintain the `BOOL8` enum name on the chance that we add a 1 bit per bool type in the future to match Arrow (which we can call `BOOL1`).

Ok, thats a good idea.

**Comment by harrism** [comment]
Moving to 0.14

File: CHANGELOG.md [hunk 1]
```diff
@@ -423,6 +423,7 @@
 - PR #3673 Parquet reader: improve rounding of timestamp conversion to seconds
 - PR #3699 Stringify libcudacxx headers for binary op JIT
 - PR #3697 Improve column insert performance for wide frames
+- PR #3581 Remove `bool8`
```
Comment (kkraus14) [inline CHANGELOG.md]: Move to 0.14 changelog section

File: cpp/tests/utilities/column_wrapper.hpp [hunk 2]
```diff
@@ -199,21 +197,18 @@ auto make_chars_and_offsets(StringsIterator begin, StringsIterator end,
  *
  * @tparam Element The fixed-width element type
  *---------------------------------------------------------------------------**/
-template <typename Element>
+template <typename ElementTo, typename ElementFrom = ElementTo>
```
Comment (jrhemstad) [inline cpp/tests/utilities/column_wrapper.hpp]: Does `ElementFrom` really need to be a part of the class's type?

Can't the init list ctor be a template? It would be ideal to avoid having to specify the explicit type param in `fixed_width_column_wrapper`. 

https://wandbox.org/permlink/oEi9LrZe1sXAERoq

Comment (codereport) [reply cpp/tests/utilities/column_wrapper.hpp]: Let me give it another shot. I also got a simplified example working in godbolt when I initially attempted, but the corresponding change did not succeed in the codebase. I will report back.

File: CHANGELOG.md (hunk 1 above)
Comment (codereport) [reply CHANGELOG.md]: Done :)

File: cpp/include/cudf/copying.hpp [hunk 3]
```diff
@@ -490,11 +490,11 @@ std::vector<contiguous_split_result> contiguous_split(cudf::table_view const& in
  *          
  * @throws cudf::logic_error if lhs and rhs are not of the same type
  * @throws cudf::logic_error if lhs and rhs are not of the same length 
- * @throws cudf::logic_error if boolean mask is not of type bool8
+ * @throws cudf::logic_error if boolean mask is not of type bool
  * @throws cudf::logic_error if boolean mask is not of the same length as lhs and rhs  
  * @param[in] left-hand column_view
  * @param[in] right-hand column_view
- * @param[in] column of `BOOL8` representing "left (true) / right (false)" boolean for each element and
+ * @param[in] column of `BOOL` representing "left (true) / right (false)" boolean for each element and
```
[the remark this answers is not in the document]
Comment (codereport) [reply cpp/include/cudf/copying.hpp]: Done

File: cpp/tests/utilities/column_wrapper.hpp [hunk 4]
```diff
@@ -199,21 +197,18 @@ auto make_chars_and_offsets(StringsIterator begin, StringsIterator end,
...
 class fixed_width_column_wrapper : public detail::column_wrapper {
-  static_assert(cudf::is_fixed_width<Element>(),
```
[the remark this answers is not in the document]
Comment (codereport) [reply cpp/tests/utilities/column_wrapper.hpp]: Oversight on my part. Must have been temporary change that I forgot to add back.

It has been added back now.

File: cpp/include/cudf/types.hpp [hunk 5]
```diff
@@ -169,7 +169,7 @@ enum type_id {
   INT64,      ///< 8 byte signed integer
   FLOAT32,    ///< 4 byte floating point
   FLOAT64,    ///< 8 byte floating point
-  BOOL8,      ///< Boolean using one byte per value, 0 == false, else true
```
Comment (harrism) [inline cpp/include/cudf/types.hpp]: Why remove documentation if the type ID is staying?

File: cpp/tests/grouped_rolling/grouped_rolling_test.cpp [hunk 6]
```diff
@@ -112,8 +112,8 @@ class GroupedRollingTest : public cudf::test::BaseFixture {
                                 size_type min_periods)
   {
     size_type num_rows = input.size();
-    std::vector<cudf::size_type> ref_data(num_rows);
-    std::vector<bool> ref_valid(num_rows);
+    thrust::host_vector<cudf::size_type> ref_data(num_rows);
```
Comment (harrism) [inline cpp/tests/grouped_rolling/grouped_rolling_test.cpp]: Why is it necessary to change to `host_vector` from `vector`?

File: cpp/tests/utilities/column_utilities.cu [hunk 7]
```diff
@@ -295,27 +295,36 @@ std::vector<bitmask_type> bitmask_to_host(cudf::column_view const& c) {
...
+        out.begin(),
+        [&h_data](auto idx) {
+          return bit_is_set(h_data.second.data(), idx) ? std::to_string(h_data.first[idx]) : std::string("NULL");
+        });
+
     } else {
-      std::transform(h_data.first.begin(), h_data.first.end(), out.begin(), [](Element el) {
+
+      std::transform(
+        h_data.first.begin(),
```
Comment (harrism) [inline cpp/tests/utilities/column_utilities.cu]: Why do this, if you don't do it for all function calls? (Which I hope we wouldn't do, since it will unnecessarily make our code much longer to scroll through.)

File: cpp/tests/utilities/column_utilities.hpp [hunk 8]
```diff
@@ -94,7 +94,7 @@ std::string to_string(cudf::column_view const& col, std::string const& delimiter
  *
  * @param col The column view
  *---------------------------------------------------------------------------**/
-std::vector<std::string> to_strings(cudf::column_view const& col);
+thrust::host_vector<std::string> to_strings(cudf::column_view const& col);
```
Comment (harrism) [inline cpp/tests/utilities/column_utilities.hpp]: This change seems pretty pervasive. I would prefer to stick with STL containers wherever possible.

File: cpp/tests/utilities/column_wrapper.hpp [hunk 9]
```diff
@@ -137,7 +137,7 @@ std::vector<bitmask_type> make_null_mask_vector(ValidityIterator begin,
       cudf::bitmask_allocation_size_bytes(size) / sizeof(bitmask_type);
   std::vector<bitmask_type> null_mask(num_words, 0);
   for (auto i = 0; i < size; ++i) {
-    if (begin[i] == true) {
+    if (*(begin + i)) {
```
Comment (harrism) [inline cpp/tests/utilities/column_wrapper.hpp]: I think the `[i]` version is clearer. Is there a reason to change this?

File: cpp/tests/utilities_tests/column_utilities_tests.cpp [hunk 10]
```diff
@@ -173,27 +175,21 @@ TEST_F(ColumnUtilitiesStringsTest, StringsToHostAllNulls)
...
-  std::ostringstream tmp;
-
-  int index = 0;
-  for (auto x : std_col) {
-    tmp << ((index == 0) ? "" : delimiter);
-    tmp << std::to_string(x);
-    ++index;
-  }
+  auto to_string = [] (auto e) { return std::to_string(e); };
+  auto const expected = boost::join(std_col | boost::adaptors::transformed(to_string), delimiter);
```
Comment (harrism) [inline cpp/tests/utilities_tests/column_utilities_tests.cpp]: libcudf is not dependent on boost, please remove this. (Yes, arrow is dependent on boost, but we have successfully avoided a direct libcudf->boost dependency up to now.)

File: cpp/tests/utilities/column_wrapper.hpp (hunk 2 above)
Comment (codereport) [reply cpp/tests/utilities/column_wrapper.hpp]: @jrhemstad A couple of the issues I am working on right now (these are the easier ones):

https://wandbox.org/permlink/iHjJQmAIQDapHAW1

Comment (codereport) [reply cpp/tests/utilities/column_wrapper.hpp]: Solution to issue 2:
```cpp
    // Issue 2
    using f = float;
    column_wrapper<float>  a{f(0.5), f(1.0), f(1.5), NAN};
    column_wrapper<double> b{0.5, 1.0, 1.5, double(NAN)};
```

File: cpp/tests/grouped_rolling/grouped_rolling_test.cpp (hunk 6 above)
Comment (jrhemstad) [reply cpp/tests/grouped_rolling/grouped_rolling_test.cpp]: [vector of bool](https://howardhinnant.github.io/onvectorbool.html)

File: cpp/tests/utilities/column_wrapper.hpp (hunk 9 above)
Comment (jrhemstad) [reply cpp/tests/utilities/column_wrapper.hpp]: Not all iterators support `operator[]`. This is definitely the more canonical way of handling iterators.

File: cpp/tests/unary/unary_ops_test.cu [hunk 11]
```diff
@@ -71,15 +71,15 @@ TYPED_TEST_CASE(cudf_logical_test, cudf::test::NumericTypes);
 TYPED_TEST(cudf_logical_test, LogicalNot) {
   cudf::size_type colSize = 1000;
   std::vector<TypeParam> h_input_v(colSize, false);
-  std::vector<cudf::experimental::bool8> h_expect_v(colSize);
+  std::vector<bool> h_expect_v(colSize);
```
Comment (aschaffer) [inline cpp/tests/unary/unary_ops_test.cu]: `thrust::host_vector<bool>` to avoid `vector<bool>` proxy mechanism?

File: cpp/src/stream_compaction/legacy/apply_boolean_mask.cu [hunk 12]
```diff
@@ -27,19 +27,19 @@ template <bool has_data, bool has_nulls>
 struct boolean_mask_filter
 {
   boolean_mask_filter(gdf_column const & boolean_mask) :
-    data{static_cast<cudf::bool8 *>(boolean_mask.data)},
+    data{static_cast<bool *>(boolean_mask.data)},
     bitmask{reinterpret_cast<bit_mask_t *>(boolean_mask.valid)}
```
[the remark this answers is not in the document]
Comment (jrhemstad) [reply cpp/src/stream_compaction/legacy/apply_boolean_mask.cu]: This is legacy code where the reinterpret cast is necessary. It's not worth spending any more time on as it's something we've already remedied in libcudf++.

[the remark this answers is not in the document]
Comment (aschaffer) [reply cpp/src/stream_compaction/legacy/apply_boolean_mask.cu]: Ok, please feel free to resolve, then.

File: cpp/tests/utilities/column_wrapper.hpp (hunk 2 above)
Comment (codereport) [reply cpp/tests/utilities/column_wrapper.hpp]: I think the fundamental problem is that currently, the `fixed_width_column_wrapper` constructors that take `initializer_list`s (and the current variant) don't use deduction. `Element` (from before) and now `ElementTo` specify the type. It means that:
```cpp
fixed_width_column_wrapper<TypeParam> col{1, 2, 3, 4.0, 5.0. 6.0};
```
will compile now and fails with the proposed change. Now, I have gone and fixed all the obvious ones that are similar to above - but there are some trickier ones that I am still trying to resolve (things to do with `timestamp`, etc). 

Will update later with progress.

File: cpp/tests/utilities_tests/column_utilities_tests.cpp (hunk 10 above)
Comment (trxcllnt) [reply cpp/tests/utilities_tests/column_utilities_tests.cpp:182]: We do use `boost::filesystem` in `jit/cache.cpp` to manage the JIT cache directory: https://github.com/rapidsai/cudf/blob/fc4de2c76b8bab91b9274f14853618302fb02ede/cpp/src/jit/cache.cpp#L46-L51

Comment (aschaffer) [reply cpp/tests/utilities_tests/column_utilities_tests.cpp:182]: Also we link against `boost` libraries (`cpp/CMakeLists.txt`):
```
target_link_libraries(cudf NVCategory NVStrings rmm ${ARROW_CUDA_LIB_LINK} ${ARROW_LIB} nvrtc cudart cuda $\
   {ZLIB_LIBRARIES} ${**Boost_LIBRARIES**})
```

We certainly have a strong dependency in the `cpp/CMakeLists.txt`: 
```
message(FATAL_ERROR "Boost not found, please check your settings.")
``` 
if `boost` not found.

File: cpp/tests/utilities/column_wrapper.hpp (hunk 2 above)
Comment (jrhemstad) [reply cpp/tests/utilities/column_wrapper.hpp:202]: Issue 1 is just a result of the fact that I used `std::vector` in my example. If you replace with `thrust::host_vector` I'd expect it to work correct. 

Issue 2 I've seen before. I can't recall exactly right now, but there's some trickiness in the type deduction that's leading to ambiguity. I know I ran into this before with the column_wrapper. I'll have to look back and see if I can't find what I figured out before.

Comment (jrhemstad) [reply cpp/tests/utilities/column_wrapper.hpp:202]: Okay, so for issue 2, the problem is that the type `U` for the `std::initializer_list<U>` is now deduced (as you said), so you have to make sure everything in the list has the same type (since conversions aren't considered during type deduction). 

You're getting bit by this because `1.0` is type `double` and `NAN` is type `float`. Since those have different types, the compiler is complaining about deducing conflicting types: https://wandbox.org/permlink/Ji4XYOL6ACPZ7xTu 

So there's a few ways to solve that other than your proposed solution, e.g., you could also do:
```
     column_wrapper<float>  a{{0.5, 1.0, 1.5, std::numeric_limits<double>::quiet_NaN()}};
     column_wrapper<double> b{0.5, 1.0, 1.5, std::numeric_limits<double>::quiet_NaN()};
```

I imagine it's uncommon to try and specify objects of different types in the init list for a column wrapper. Are you finding this to be a common problem?

File: cpp/include/cudf/utilities/traits.hpp [hunk 13]
```diff
@@ -105,16 +105,13 @@ constexpr inline bool is_numeric(data_type type) {
...
  * @brief Indicates whether `T` is a Boolean type.
  *
- * `cudf::bool8` is cudf's Boolean type.
- *
  * @param type The `data_type` to verify
  * @return true `type` is Boolean
  * @return false `type` is not Boolean
  *---------------------------------------------------------------------------**/
 template <typename T>
 constexpr inline bool is_boolean() {
```
[the remark this answers is not in the document]
Comment (codereport) [reply cpp/include/cudf/utilities/traits.hpp:113]: Arguably no, but many of these functions seem to just wrap a `std::is_same`, and it is used in ~30 places throughout our code base. Seems like we should leave it for consistency.

File: cpp/include/cudf/utilities/traits.hpp [hunk 14]
```diff
@@ -105,16 +105,13 @@ constexpr inline bool is_numeric(data_type type) {
...
-  return std::is_same<T, cudf::experimental::bool8>::value ||
-         std::is_same<T, bool>::value;
+  return std::is_same<T, bool>::value;
 }
 
 struct is_boolean_impl {
```
[the remark this answers is not in the document]
Comment (codereport) [reply cpp/include/cudf/utilities/traits.hpp:117]: Yes, done.

File: cpp/include/cudf/types.hpp (hunk 5 above)
Comment (codereport) [reply cpp/include/cudf/types.hpp:172]: This was an oversight. I deleted it at one point and when I added back clearly forgot the comment. Fixed.

File: cpp/tests/grouped_rolling/grouped_rolling_test.cpp (hunk 6 above)
Comment (codereport) [reply cpp/tests/grouped_rolling/grouped_rolling_test.cpp:115]: Yes, this caused me a lot of headaches. Probably one of C++ biggest mistakes.

File: cpp/tests/utilities/column_wrapper.hpp (hunk 9 above)
Comment (codereport) [reply cpp/tests/utilities/column_wrapper.hpp:140]: IIRC I made this change specifically because of the fact that it was necessary.

File: cpp/tests/utilities/column_utilities.cu (hunk 7 above)
Comment (codereport) [reply cpp/tests/utilities/column_utilities.cu:316]: This just comes down to preference. I think it makes the code much more readable, arguably even beautiful - and there are many cases where this is they style being used. A really good example is in @karthikeyann recent Rank PR, in the file:

https://github.com/rapidsai/cudf/blob/branch-0.14/cpp/src/sort/rank.cu

Some of this code is beautiful/so much more readable because of the vertical formatting:

```cpp
template<typename outputType>
void rank_min(rmm::device_vector<size_type> const &group_keys,
              column_view sorted_order_view, mutable_column_view rank_mutable_view, cudaStream_t stream) {
  // min of first in the group
  // All equal values have min of ranks among them.
  // algorithm: reduce_by_key(dense_rank, 1, n, min), scatter
  tie_break_ranks_transform<size_type>(
      group_keys,
      thrust::make_counting_iterator<size_type>(1),
      sorted_order_view,
      rank_mutable_view.begin<outputType>(),
      thrust::minimum<size_type>{},
      thrust::identity<outputType>{},
      stream);
}

template<typename outputType>
void rank_max(rmm::device_vector<size_type> const &group_keys,
              column_view sorted_order_view, mutable_column_view rank_mutable_view, cudaStream_t stream) {
  // max of first in the group
  // All equal values have max of ranks among them.
  // algorithm: reduce_by_key(dense_rank, 1, n, max), scatter
  tie_break_ranks_transform<size_type>(
      group_keys,
      thrust::make_counting_iterator<size_type>(1),
      sorted_order_view,
      rank_mutable_view.begin<outputType>(),
      thrust::maximum<size_type>{},
      thrust::identity<outputType>{},
      stream);
}

// ...

    thrust::transform(
        rmm::exec_policy(stream)->on(stream),
        rank_iter,
        rank_iter + input.size(),
        rank_iter,
        [is_dense, drs, count] __device__(double r) -> double {
          return is_dense ? r / drs[count - 1] : r / count;
        });
```
As Walter Brown said in his Meeting C++ 2019 keynote: 
**Most people don't take advantage of their eyes**
https://youtu.be/OAmWHmwlMwI?t=3195

All this being said, for this case it it doesn't have as a big a trade-off when it comes to readability so I can switch it back

File: cpp/tests/io/csv_test.cu [hunk 15]
```diff
@@ -206,10 +206,10 @@ TEST_F(CsvReaderTest, Booleans) {
   expect_column_data_equal(std::vector<int32_t>{1, 0, 0, 0, 1}, view.column(0));
   expect_column_data_equal(std::vector<int16_t>{0, 1, 1, 0, 1}, view.column(2));
   expect_column_data_equal(
-      std::vector<cudf::experimental::bool8>{
-          cudf::experimental::true_v, cudf::experimental::true_v,
-          cudf::experimental::false_v, cudf::experimental::true_v,
-          cudf::experimental::false_v},
+      std::vector<bool>{
```
[the remark this answers is not in the document]
Comment (codereport) [reply cpp/tests/io/csv_test.cu:209]: We can't use `thrust::host_vector` here because it doesn't support initializer_list constructor.

[the remark this answers is not in the document]
Comment (codereport) [reply cpp/tests/io/csv_test.cu:209]: @jrhemstad Are we saying we shouldn't use `std::vector<bool>` anywhere? It is pervasively used in our testing and I assumed it was OK pending your use case doesn't run into issues.

[the remark this answers is not in the document]
Comment (jrhemstad) [reply cpp/tests/io/csv_test.cu:209]: > I assumed it was OK pending your use case doesn't run into issues.

Yeah, as long as its being used in a way that doesn't cause issues, I see no problem with it.

[the remark this answers is not in the document]
Comment (aschaffer) [reply cpp/tests/io/csv_test.cu:209]: OK, then I will assume all the occurrences of `vector<bool>` that I flagged below as possibly problematic are instances of the same use case. In which you can just resolve each such conversation.

[the remark this answers is not in the document]
Comment (codereport) [reply cpp/tests/io/csv_test.cu:209]: Great - yea the use cases I did change were the ones that were problematic.

File: cpp/tests/utilities/column_utilities.hpp (hunk 8 above)
Comment (codereport) [reply cpp/tests/utilities/column_utilities.hpp:97]: See comment above. But basically everywhere I use `thrust::host_vector` is necessary. I do agree that using STL containers as much as possible is preferable and that is what we are dong :)

File: cpp/tests/utilities/column_wrapper.hpp (hunk 2 above)
Comment (codereport) [reply cpp/tests/utilities/column_wrapper.hpp:202]: Yes! I have it working. So it turns out the FINAL issue was the wonderful `{ }` vs `( )` construction. Now that the initializer list isn't told the type and has to deduce, if can't resolve the correct construction call with braces `{ }`, so every instance that uses them has to be replaced with this `( )`. It took me a while to figure out because the way that the problem manifested itself via compiler error message was super misleading.

Anyway, will push code tomorrow after cleaning it up.

File: cpp/tests/utilities_tests/column_utilities_tests.cpp (hunk 10 above)
Comment (harrism) [reply cpp/tests/utilities_tests/column_utilities_tests.cpp:182]: Hmmm, didn't realize that. The boost::filesystem dependency will be removed when we can move up to C++17. I don't think this use of boost really improves the code. In fact, since I've never used those features of boost, I don't know what the new code does. And since it's not `std::`, it's not trivial to look up.  I would stick to std:: here.

File: cpp/include/cudf/utilities/traits.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/include/cudf/utilities/traits.hpp:113]: OK, if this version of is_boolean() is used, then it should stay. I was unsure if it was not used since it is called by is_boolean(data_type).

File: cpp/tests/grouped_rolling/grouped_rolling_test.cpp (hunk 6 above)
Comment (harrism) [reply cpp/tests/grouped_rolling/grouped_rolling_test.cpp:115]: Right, I remember the discussion now. We'll have to watch out for this in new tests that are added.

File: cpp/tests/utilities/column_wrapper.hpp (hunk 2 above)
Comment (jrhemstad) [reply cpp/tests/utilities/column_wrapper.hpp:202]: I've been bitten by that brace init list issue several times with `column_wrapper`. The compiler error really is super unhelpful.

Comment (codereport) [reply cpp/tests/utilities/column_wrapper.hpp]: Changes have been pushed :)

File: cpp/tests/utilities_tests/column_utilities_tests.cpp (hunk 10 above)
Comment (codereport) [reply cpp/tests/utilities_tests/column_utilities_tests.cpp]: As I posted on Slack, unfortunately, this code:
```cpp
  std::stringstream tmp;

  std::copy(
    std::begin(std_col),
    std::prev(std::end(std_col)),
    std::ostream_iterator<TypeParam>(tmp, delimiter));

  tmp << std::to_string(std_col.back());
```
Doesn't work for all types. It fails for `float`, `double` and `signed char`. You would first need to call `std::transform` on a temporary `vector` or use a transform iterator.
I have changed it back to the old code.

File: cpp/src/io/json/json_gpu.cu [hunk 16]
```diff
@@ -126,26 +126,6 @@ __inline__ __device__ T decode_value(const char *data, long start, long end,
   return cudf::experimental::io::gpu::parse_numeric<T>(data, start, end, opts);
 }
 
-
```
[the remark this answers is not in the document]
Comment (codereport) [reply cpp/src/io/json/json_gpu.cu:129]: Hmmm, this might have been an oversight. I think with all the deleting of code I did, I did a search for where this is used - and it wasn't used anywhere - but it just seems like in general the `decode_value` functions aren't used / tested anywhere. Seems like we will need it. I will add versions for `bool`

File: cpp/tests/utilities_tests/column_utilities_tests.cpp (hunk 10 above)
Comment (harrism) [reply cpp/tests/utilities_tests/column_utilities_tests.cpp]: What's wrong with a transform iterator?

Comment (codereport) [reply cpp/tests/utilities_tests/column_utilities_tests.cpp]: Nothing, I mentioned it in the Slack thread. I switched it to the `std::copy` + a `thrust::transform_iterator`

Comment (codereport) [reply cpp/tests/utilities_tests/column_utilities_tests.cpp]: Nothing, I mentioned it in the Slack thread. I switched it to the `std::copy` + a `thrust::transform_iterator`

File: python/cudf/cudf/_lib/cpp/wrappers/bool.pxd [hunk 17]
```diff
@@ -1,6 +0,0 @@
-# Copyright (c) 2020, NVIDIA CORPORATION.
```
Comment (shwina) [inline python/cudf/cudf/_lib/cpp/wrappers/bool.pxd:1]: 🔥

Comment (shwina) [reply python/cudf/cudf/_lib/cpp/wrappers/bool.pxd:1]: Deleted code is the best kind of code

File: cpp/tests/utilities/type_lists.hpp [hunk 18]
```diff
@@ -61,8 +60,19 @@ constexpr auto types_to_ids() {
   constexpr auto N = GetSize<TYPES>;
   return types_to_ids_impl<TYPES>(std::make_index_sequence<N>());
 }
+
 }  // namespace detail
 
+template <typename TypeParam, typename T>
+auto make_type_param_vector(std::initializer_list<T> const& init_list) {
```
Comment (karthikeyann) [inline cpp/tests/utilities/type_lists.hpp:67]: Why is this necessary?
I suppose, one reason is narrowing conversions limitations. 
https://en.cppreference.com/w/cpp/language/list_initialization#Narrowing_conversions

File: cpp/tests/replace/clamp_test.cpp [hunk 19]
```diff
@@ -262,7 +262,7 @@ TYPED_TEST(ClampFloatTest, WithNANandNoNull) {
     using T = TypeParam;
     using ScalarType = cudf::experimental::scalar_type_t<T>;
 
-    cudf::test::fixed_width_column_wrapper<T> input({8.0, 6.0, NAN, 3.0, 4.0, 5.0, 1.0, NAN, 2.0, 9.0});
+    cudf::test::fixed_width_column_wrapper<T> input({T(8.0), T(6.0), T(NAN), T(3.0), T(4.0), T(5.0), T(1.0), T(NAN), T(2.0), T(9.0)});
```
Comment (karthikeyann) [inline cpp/tests/replace/clamp_test.cpp:265]: Too verbose `T()` for every value. Any less verbose alternative?

Comment (codereport) [reply cpp/tests/replace/clamp_test.cpp:265]: It is a lot less verbose than `static_cast<T>()`

File: cpp/tests/utilities/type_lists.hpp (hunk 18 above)
Comment (codereport) [reply cpp/tests/utilities/type_lists.hpp:67]: Yea, we need this for casting to `bool`.
